### PR TITLE
feat: Added slack chat integration.

### PR DESCRIPTION
**Pull request by osala-eng**
## Description

Slack chat bot integration

Implements #227 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor and code improvement (non-breaking change which improves code quality and/or performance)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation
- [ ] Tests
- [ ] Other chores such as maintenance

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration by modifying the list below.

- [ ] Test A
- [ ] Test B
- [ ] Test C

**Test Configuration**:

Please describe the test setup. List them below as bullet points.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings
- [ ] The commit message follows the convention of this project

### CODE REVIEW

**Comment by osala-eng** [comment]
@janaka  I've setup slack chat bot. you can review this and tell me what you think.

I've also added the sample chat bot to the Docq.AI's slack work space you can try that out aswell.

**Comment by janaka** [comment]
Great. I'll take a look.

**Comment by janaka** [comment]
How do I do RAG? Or you waiting for the updated Web API live?

**Comment by osala-eng** [comment]
> How do I do RAG? Or you waiting for the updated Web API live?

This just implements chat comletion as described in #227, once we get this approved i'll use it to implement Retrival in #228

**Comment by janaka** [comment]
great, I'll take a look at the code this weekend. 

I shipped the new Web API btw.

**Review by janaka — COMMENTED** [review]
Looks like the whole integration is webhooks based? 

This code should be part of the API I think. It also should be designed as a general webhook handler ideally.

File: source/docq/integrations/slack.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    team_id TEXT NOT NULL,
+    team_name TEXT NOT NULL,
+    org_id INTEGER NOT NULL,
+    space_group_id INTEGER,
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    FOREIGN KEY (org_id) REFERENCES orgs(id),
+    FOREIGN KEY (space_group_id) REFERENCES space_groups(id),
+    UNIQUE (app_id, team_id, org_id)
+);
+"""
```
Comment (janaka) [inline source/docq/integrations/slack.py:26]: Can you explain the data in this table pls. it's not obvious to me.

I don't love having Slack specific tables. It's OK that you started like this to get the whole thing up and running end to end. 

I'm wondering if we should iterate to a similar/same pattern as config for data sources.

File: source/docq/setup.py [hunk 2]
```diff
@@ -38,6 +39,7 @@ def init() -> None:
         manage_assistants._init()
         db_migrations.run() # run db migrations after all tables are created
         services._init()
+        slack._init()
```
Comment (janaka) [inline source/docq/setup.py:42]: pls put this inside and integrations._init() and call that here.

File: source/docq/support/auth_utils.py [hunk 3]
```diff
@@ -45,15 +45,18 @@ def init_session_cache() -> None:
         raise ValueError("DOCQ_COOKIE_HMAC_SECRET_KEY must be 32 or more characters")
 
 
-def _set_cookie(cookie: str) -> None:
+def _set_cookie(cookie: str, name: Optional[str] = None, expiry: Optional[datetime] = None) -> None:
     """Set client cookie for authentication."""
-    try:
+    if name is None:
+        name = SESSION_COOKIE_NAME
```
Comment (janaka) [inline source/docq/support/auth_utils.py:51]: this shouldn't be here now right? That value should be passed in.

and name should not be Optional therefore.

File: source/docq/support/auth_utils.py [hunk 4]
```diff
@@ -45,15 +45,18 @@ def init_session_cache() -> None:
...
+    if expiry is None:
```
Comment (janaka) [inline source/docq/support/auth_utils.py:52]: thinking it makes more sense for this function param to be`ttl` with the default  set to `TTL_HOURS`. Also renamed to COOKIE_DEFAULT_TTL_HOURS.

Comment (janaka) [reply source/docq/support/auth_utils.py:52]: unless I'm missing a use case.

File: web/api/integration/slack/slack_application.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        database=get_sqlite_shared_system_file(),
+        install_path="/api/integration/slack/v1/install",
+        redirect_uri_path="/api/integration/slack/v1/oauth_redirect",
+        client_id=CLIENT_ID,
+        client_secret=CLIENT_SECRET,
+        scopes=SCOPES,
+        user_scopes=USER_SCOPES,
+        callback_options=CallbackOptions(success=success_callback, failure=failure_callback),
+    )
+)
```
Comment (janaka) [inline web/api/integration/slack/slack_application.py:40]: what's going on here? this seems off with sqlite3 being called directly. Database calls should all be contained strictly within the data layer.

File: web/api/integration/slack/oauth_flow.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+from slack_sdk.oauth.installation_store.sqlite3 import SQLite3InstallationStore
+from slack_sdk.oauth.state_store.sqlite3 import SQLite3OAuthStateStore
+from slack_sdk.web import WebClient
+
+
+class SlackOAuthFlow(OAuthFlow):
+    """Custom slack oauth flow."""
+
+    @classmethod
+    def sqlite3(
```
Comment (janaka) [inline web/api/integration/slack/oauth_flow.py:25]: same comment here re: sqlite

File: web/api/integration/slack/chat_handler.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+"""Slack chat action handler."""
```
Comment (janaka) [inline web/api/integration/slack/chat_handler.py:1]: is `chat` a category of Slack events? so the pattern is we'll create a separate hanlder file to handle each category?

Comment (janaka) [reply web/api/integration/slack/chat_handler.py:1]: Generally worth referencing links to Slack docs.

**Review by janaka — CHANGES_REQUESTED** [review]
See comments. Please respond inline. 

I've not paid attention to the integrations setting / install stuff because it seems you are still working on the changes we discussed, moving those to the Admin Section.

File: source/docq/integrations/slack.py (hunk 1 above)
Comment (osala-eng) [reply source/docq/integrations/slack.py]: Added comments on this.

File: source/docq/setup.py (hunk 2 above)
Comment (osala-eng) [reply source/docq/setup.py]: Done

File: source/docq/support/auth_utils.py (hunk 3 above)
Comment (osala-eng) [reply source/docq/support/auth_utils.py]: Created a separate method for this.

File: source/docq/support/auth_utils.py (hunk 4 above)
Comment (osala-eng) [reply source/docq/support/auth_utils.py]: Done.

File: web/api/integration/slack/slack_application.py (hunk 5 above)
Comment (osala-eng) [reply web/api/integration/slack/slack_application.py]: Moved this to data layer.

**Comment by osala-eng** [comment]
@janaka I've updated the changes as requested and implemented end to end slack chat with docq bot using RAG.

image

**Comment by janaka** [comment]
Great. Will take a look tomorrow.

**Comment by janaka** [comment]
@osala-eng sorry, I merged the Llama Index update last night. Can you rebase and check all works. Mainly with the bits that bit the Web API

**Comment by osala-eng** [comment]
> @osala-eng sorry, I merged the Llama Index update last night. Can you rebase and check all works. Mainly with the bits that bit the Web API

Okay, let me rebase and test this.

**Comment by janaka** [comment]
great - do you have one or two tests for any of the new functionality build
for Slack and Chrome. Would be handy to know there's something.

**Comment by osala-eng** [comment]
@janaka There is a problem with [main](https://github.com/docqai/docq/tree/main) - I'll rebase this branch once that is sorted.

**Comment by janaka** [comment]
What's the issue on main branch? Seem to be ok, test passed etc.

**Comment by janaka** [comment]
Was that to do with your local virtual environment?

**Comment by osala-eng** [comment]
Yes, it was my local env.

**Comment by osala-eng** [comment]
This is ready for you now.

**Review by janaka — COMMENTED** [review]
@osala-eng I pushed a bunch of refactors. Any chance you can do a little quick test.

**Comment by osala-eng** [comment]
> @osala-eng I pushed a bunch of refactors. Any chance you can do a little quick test.

Okay let me do a quick test and get back to you.

File: source/docq/integrations/slack/slack_oauth_flow.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    #     return None
+
+    # def save_docq_slack_installation(self: Self, request: BoltRequest, installation: Installation) -> None:
+    #     """Save a Docq slack installation."""
+    #     docq_slack_app_state =  self.get_cookie("docq_slack_app_state", request.headers.get("cookie"))
+    #     if docq_slack_app_state is not None:
+    #         _, selected_org_id = docq_slack_app_state.split(":")
+    #         create_docq_slack_installation(installation, int(selected_org_id))
+    #     else:
+    #         raise BoltError("Login to Docq before installing the slack app.")
```
Comment (osala-eng) [inline source/docq/integrations/slack/slack_oauth_flow.py:102]: @janaka I see you've commented out some of the code.

**Review by osala-eng — COMMENTED** [review]
This dosent work fully as some of the functionality has been disabled.

File: source/docq/integrations/slack/slack_oauth_flow.py (hunk 8 above)
Comment (janaka) [reply source/docq/integrations/slack/slack_oauth_flow.py:102]: Seemed like those weren't being called. Maybe my VS Code search wasn't working properly. If you uncomment that does it work.

Comment (janaka) [reply source/docq/integrations/slack/slack_oauth_flow.py]: great! mind adding a few comments to document that logic.

Assume all is working as expected at this point?

Comment (osala-eng) [reply source/docq/integrations/slack/slack_oauth_flow.py]: Yes, I've tested and all is working as expected.

Comment (osala-eng) [reply source/docq/integrations/slack/slack_oauth_flow.py]: Let me add a few comments on this.

**Comment by osala-eng** [comment]
@janaka This is now ready.

**Comment by janaka** [comment]
@osala-eng did you write down anything / ref a link on how the values for the following were generated?


SLACK_CLIENT_ID
SLACK_CLIENT_SECRET
SLACK_SIGNING_SECRET

File: source/docq/integrations/slack/slack_application.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+from docq.integrations.slack.slack_oauth_flow import SlackOAuthFlow
+from docq.support.store import get_sqlite_shared_system_file
+from slack_bolt import App, BoltResponse
+from slack_bolt.oauth.callback_options import CallbackOptions, FailureArgs, SuccessArgs
+
+CLIENT_ID = os.environ.get("SLACK_CLIENT_ID")
+CLIENT_SECRET = os.environ.get("SLACK_CLIENT_SECRET")
+SCOPES = ["app_mentions:read", "im:history", "chat:write", "channels:read", "groups:read", "im:read", "mpim:read"]
+USER_SCOPES = ["admin"]
```
Comment (janaka) [inline source/docq/integrations/slack/slack_application.py:13]: @osala-eng wondering if we can have a lesser scope here.

**Comment by janaka** [comment]
> @osala-eng did you write down anything / ref a link on how the values for the following were generated?
> 
> SLACK_CLIENT_ID SLACK_CLIENT_SECRET SLACK_SIGNING_SECRET

@osala-eng I just remembered about the other repo with the manifest file which answer my original question sufficiently.

New related question about the env var names. You went with the SDK built-in names rather than our usual pattern of prefixing with "DOCQ_". Was that intentional for reasons like it's the only way to use the SDK? I wanted to check before refactoring to save myself some pain :)

File: source/docq/integrations/slack/slack_application.py (hunk 9 above)
Comment (osala-eng) [reply source/docq/integrations/slack/slack_application.py:16]: The only scope that you can remove without affecting the functionality is `admin` but then anyone within a slack team will be able to install the application. Currently, only admins can install the application.

For the rest, we are mostly requesting read access, except for chat, where we need write access to enable the bot to respond to mentions.

**Comment by osala-eng** [comment]
> > @osala-eng did you write down anything / ref a link on how the values for the following were generated?
> > SLACK_CLIENT_ID SLACK_CLIENT_SECRET SLACK_SIGNING_SECRET
> 
> @osala-eng I just remembered about the other repo with the manifest file which answer my original question sufficiently.
> 
> New related question about the env var names. You went with the SDK built-in names rather than our usual pattern of prefixing with "DOCQ_". Was that intentional for reasons like it's the only way to use the SDK? I wanted to check before refactoring to save myself some pain :)

Yes, you can customize the env var names, then pass them explicitly when instanciating the `SlackOauthFlow` class. By default, slack_sdk looks for the names provided in their documentation.

**Review by janaka — APPROVED** [review]

File: source/docq/integrations/slack/slack_application.py (hunk 9 above)
Comment (janaka) [reply source/docq/integrations/slack/slack_application.py]: I've removed the user_scopes `admin` for now. As I understood from the docs user scopes allow the bot to do things on behalf of a user / as a user with those perms. maybe I've misunderstood. Will revisit.

**Pull request merged**